### ISSUE TITLE
Lack of KEM algorithm testing

### ISSUE THREAD

**Issue by paulidale** [openssl/openssl#26032]
#25938 added algorithm KEM tests for ML-KEM and there are existing tests for RSA.  None of the other KEM algorithms seem to have evp_test tests.  They really ought to.

**Comment by baentsch** [comment]
Probably ignorant question: Are handshake tests part of what you refer to as "evp_test tests", @paulidale ? Those really are missing (and also would have detected problems with the new ML-KEM provider). Looking (only?) at test/evp_test.c I don't see that -- and also lots of "hardcoded" stuff -- wouldn't it be nicer to iterate through all KEMs (as per `openssl list -kem-algorithms`) and do a foreach test (handshakes included)? Or does something like this already exist somewhere?

**Comment by paulidale** [comment]
No, they aren't.  `evp_test` runs a series of known answer tests against a variety of algorithms.  The tests are contained in files in the `test/recipes/30-test_evp_data` directory.

I think iterating through all KEMs doing a handshake test would be good to do _in addition_ to the `evp_test` tests.  The former is a pairwise consistency test which, while useful, doesn't verify that the algorithm works correctly (only that the implementation can decrypt what _it_ encrypts).  The TLS tests should exercise these algorithms end to end but an independent test is likely to be worthwhile too.  I don't remember such being added.  `test/recipes/20-test_pkeyutl.t` does do some KEM testing using RSA via the command line `pkeyutl` tool.

I've added minimal ECX known answer tests in #25990 (which will merge today if the conflicts aren't too bad).

**Comment by baentsch** [comment]
IMO can be closed. Agreed, @paulidale ?

**Comment by paulidale** [comment]
I'm not sure we want to close this.  This highlights a gap in our testing via evp_test.

**Comment by paulidale** [comment]
This is done via #26946.  Closing.
